File: ios/Godstone/Sources/GodstoneMesh/BleRoleBindingCoordinator.swift
import Foundation

/// Events processed by the role-binding coordinator.
public enum BleRoleBindingEvent: Sendable {
    case discovered(UUID)
    case provisionalConnected(UUID)
    case remoteLinkInfoRead(UUID, BleLinkInfoV1)
    case remoteLinkInfoReadRaw(UUID, Data)
    case localLinkInfoWriteAcknowledged(UUID, Data)
    case incomingCentralLinkInfoWrite(UUID, Data)
    case disconnected(UUID)
    case failed(UUID, String)
}

/// Actions produced by the pure role-binding coordinator.
public enum BleRoleBindingAction: Equatable, Sendable {
    case connectProvisionally(UUID)
    case readRemoteLinkInfo(UUID)
    case writeLocalLinkInfo(UUID, remoteHint: Data)
    case roleBound(UUID, role: BleRole, remoteHint: Data)
    case cancelWrongDirectionLink(UUID, reason: String)
    case rejectIncomingWrite(UUID, reason: String)
    case acceptIncomingWrite(UUID, remoteHint: Data)
    case reset(UUID)
}

/// Pure, testable role-binding coordinator for Connect-First / Elect-Before-Handshake BLE links on iOS (ADR-002, Phase C8.4D1-R2).
public final class BleRoleBindingCoordinator: Sendable {

    public let localHint: Data

    public init(localHint: Data) {
        precondition(localHint.count == BleRoleElection.nodeHintBytes, "localHint must be exactly 4 bytes")
        self.localHint = localHint
    }

    /// Process an event on an outgoing Central connection path.
    public func processCentralEvent(_ event: BleRoleBindingEvent) -> BleRoleBindingAction {
        switch event {
        case .discovered(let peerId):
            return .connectProvisionally(peerId)
        case .provisionalConnected(let peerId):
            return .readRemoteLinkInfo(peerId)
        case .remoteLinkInfoReadRaw(let peerId, let data):
            guard let linkInfo = BleLinkInfoCodec.decode(data) else {
                return .cancelWrongDirectionLink(peerId, reason: "Malformed remote LinkInfo payload")
            }
            return evaluateCentralElection(peerId: peerId, remoteLinkInfo: linkInfo)
        case .remoteLinkInfoRead(let peerId, let linkInfo):
            return evaluateCentralElection(peerId: peerId, remoteLinkInfo: linkInfo)
        case .localLinkInfoWriteAcknowledged(let peerId, let remoteHint):
            return .roleBound(peerId, role: .initiator, remoteHint: remoteHint)
        case .disconnected(let peerId), .failed(let peerId, _):
            return .reset(peerId)
        case .incomingCentralLinkInfoWrite(let peerId, let data):
            return processPeripheralLinkInfoWrite(peerId: peerId, rawBytes: data)
        }
    }

    private func evaluateCentralElection(peerId: UUID, remoteLinkInfo: BleLinkInfoV1) -> BleRoleBindingAction {
        let election = BleRoleElection.elect(localHint: localHint, remoteHint: remoteLinkInfo.nodeHint)
        switch election {
        case .elected(let role):
            if role == .initiator {
                // local < remote: we are initiator -> write local LinkInfo with response
                return .writeLocalLinkInfo(peerId, remoteHint: remoteLinkInfo.nodeHint)
            } else {
                // local > remote: we are responder -> wrong direction, cancel central link
                return .cancelWrongDirectionLink(
                    peerId,
                    reason: "Elected RESPONDER on central link; cancelling wrong-direction connection"
                )
            }
        case .tie:
            return .cancelWrongDirectionLink(peerId, reason: "Equal node hints (tie detected); fail closed")
        case .invalid(let reason):
            return .cancelWrongDirectionLink(peerId, reason: "Invalid election: \(reason)")
        }
    }

    /// Process an incoming LinkInfo write on the Peripheral GATT server.
    public func processPeripheralLinkInfoWrite(peerId: UUID, rawBytes: Data) -> BleRoleBindingAction {
        guard let linkInfo = BleLinkInfoCodec.decode(rawBytes) else {
            return .rejectIncomingWrite(peerId, reason: "Malformed LinkInfo payload")
        }

        let election = BleRoleElection.elect(localHint: localHint, remoteHint: linkInfo.nodeHint)
        switch election {
        case .elected(let role):
            if role == .responder {
                // remote < local: Central is valid initiator, we are RESPONDER -> accept and bind role
                return .acceptIncomingWrite(peerId, remoteHint: linkInfo.nodeHint)
            } else {
                // remote > local: Central should not initiate -> reject write
                return .rejectIncomingWrite(peerId, reason: "Remote hint >= local hint; rejected")
            }
        case .tie:
            return .rejectIncomingWrite(peerId, reason: "Tie detected")
        case .invalid(let reason):
            return .rejectIncomingWrite(peerId, reason: reason)
        }
    }
}

File: android/mesh/src/main/java/io/godstone/mesh/transport/BleRoleBindingCoordinator.kt
package io.godstone.mesh.transport

/**
 * Pure, testable role-binding coordinator for Connect-First / Elect-Before-Handshake BLE links (ADR-002, Phase C8.4D1-R2).
 *
 * Owns no platform Bluetooth state, no coroutines, no SessionManager, and no cryptographic keys.
 * Encapsulates the deterministic role election and LinkInfo exchange decisions.
 */
sealed interface BleRoleBindingEvent {
    data class Discovered(val peerAddress: String) : BleRoleBindingEvent
    data class ProvisionalConnected(val peerAddress: String) : BleRoleBindingEvent
    data class RemoteLinkInfoRead(val peerAddress: String, val linkInfo: BleLinkInfoV1) : BleRoleBindingEvent
    data class RemoteLinkInfoReadRaw(val peerAddress: String, val bytes: ByteArray) : BleRoleBindingEvent
    data class LocalLinkInfoWriteAcknowledged(val peerAddress: String, val remoteHint: ByteArray) : BleRoleBindingEvent
    data class IncomingCentralLinkInfoWrite(val peerAddress: String, val bytes: ByteArray) : BleRoleBindingEvent
    data class Disconnected(val peerAddress: String) : BleRoleBindingEvent
    data class Failed(val peerAddress: String, val reason: String) : BleRoleBindingEvent
}

sealed interface BleRoleBindingAction {
    data class ConnectProvisionally(val peerAddress: String) : BleRoleBindingAction
    data class ReadRemoteLinkInfo(val peerAddress: String) : BleRoleBindingAction
    data class WriteLocalLinkInfo(val peerAddress: String, val remoteHint: ByteArray) : BleRoleBindingAction
    data class RoleBound(val peerAddress: String, val role: BleRole, val remoteHint: ByteArray) : BleRoleBindingAction
    data class CancelWrongDirectionLink(val peerAddress: String, val reason: String) : BleRoleBindingAction
    data class RejectIncomingWrite(val peerAddress: String, val reason: String) : BleRoleBindingAction
    data class AcceptIncomingWrite(val peerAddress: String, val remoteHint: ByteArray) : BleRoleBindingAction
    data class Reset(val peerAddress: String) : BleRoleBindingAction
}

class BleRoleBindingCoordinator(
    val localHint: ByteArray
) {
    init {
        require(localHint.size == BleRoleElection.NODE_HINT_BYTES) {
            "localHint must be exactly ${BleRoleElection.NODE_HINT_BYTES} bytes, got ${localHint.size}"
        }
    }

    /**
     * Process an event on an outgoing Central connection path.
     */
    fun processCentralEvent(event: BleRoleBindingEvent): BleRoleBindingAction {
        return when (event) {
            is BleRoleBindingEvent.Discovered -> {
                BleRoleBindingAction.ConnectProvisionally(event.peerAddress)
            }
            is BleRoleBindingEvent.ProvisionalConnected -> {
                BleRoleBindingAction.ReadRemoteLinkInfo(event.peerAddress)
            }
            is BleRoleBindingEvent.RemoteLinkInfoReadRaw -> {
                val linkInfo = BleLinkInfoCodec.decode(event.bytes)
                if (linkInfo == null) {
                    BleRoleBindingAction.CancelWrongDirectionLink(event.peerAddress, "Malformed remote LinkInfo payload")
                } else {
                    evaluateCentralElection(event.peerAddress, linkInfo)
                }
            }
            is BleRoleBindingEvent.RemoteLinkInfoRead -> {
                evaluateCentralElection(event.peerAddress, event.linkInfo)
            }
            is BleRoleBindingEvent.LocalLinkInfoWriteAcknowledged -> {
                BleRoleBindingAction.RoleBound(event.peerAddress, BleRole.INITIATOR, event.remoteHint)
            }
            is BleRoleBindingEvent.Disconnected, is BleRoleBindingEvent.Failed -> {
                val addr = when (event) {
                    is BleRoleBindingEvent.Disconnected -> event.peerAddress
                    is BleRoleBindingEvent.Failed -> event.peerAddress
                    else -> ""
                }
                BleRoleBindingAction.Reset(addr)
            }
            is BleRoleBindingEvent.IncomingCentralLinkInfoWrite -> {
                processPeripheralLinkInfoWrite(event.peerAddress, event.bytes)
            }
        }
    }

    private fun evaluateCentralElection(peerAddress: String, remoteLinkInfo: BleLinkInfoV1): BleRoleBindingAction {
        val election = BleRoleElection.elect(localHint, remoteLinkInfo.nodeHint)
        return when (election) {
            is BleRoleElectionResult.Elected -> {
                if (election.role == BleRole.INITIATOR) {
                    // local < remote: we are the initiator -> write local LinkInfo with response
                    BleRoleBindingAction.WriteLocalLinkInfo(peerAddress, remoteLinkInfo.nodeHint)
                } else {
                    // local > remote: we are the responder -> wrong direction, cancel central link immediately
                    BleRoleBindingAction.CancelWrongDirectionLink(
                        peerAddress,
                        "Elected RESPONDER on central link; cancelling wrong-direction connection"
                    )
                }
            }
            BleRoleElectionResult.Tie -> {
                BleRoleBindingAction.CancelWrongDirectionLink(peerAddress, "Equal node hints (tie detected); fail closed")
            }
            is BleRoleElectionResult.Invalid -> {
                BleRoleBindingAction.CancelWrongDirectionLink(peerAddress, "Invalid election: ${election.reason}")
            }
        }
    }

    /**
     * Process an incoming LinkInfo write on the Peripheral GATT server.
     */
    fun processPeripheralLinkInfoWrite(peerAddress: String, rawBytes: ByteArray): BleRoleBindingAction {
        val linkInfo = BleLinkInfoCodec.decode(rawBytes)
            ?: return BleRoleBindingAction.RejectIncomingWrite(peerAddress, "Malformed LinkInfo payload")

        val election = BleRoleElection.elect(localHint, linkInfo.nodeHint)
        return when (election) {
            is BleRoleElectionResult.Elected -> {
                if (election.role == BleRole.RESPONDER) {
                    // remote < local: Central is valid initiator, we are RESPONDER -> accept write and bind role
                    BleRoleBindingAction.AcceptIncomingWrite(peerAddress, linkInfo.nodeHint)
                } else {
                    // remote > local: Central should not initiate -> reject write
                    BleRoleBindingAction.RejectIncomingWrite(peerAddress, "Remote hint >= local hint; rejected")
                }
            }
            BleRoleElectionResult.Tie -> BleRoleBindingAction.RejectIncomingWrite(peerAddress, "Tie detected")
            is BleRoleElectionResult.Invalid -> BleRoleBindingAction.RejectIncomingWrite(peerAddress, election.reason)
        }
    }
}
